### ISSUE TITLE
Too large parquet files via "COPY TO"

### ISSUE THREAD

**Issue by mayer79** [duckdb/duckdb#3316]
#### What happens?
Using "COPY TO" query produces much larger parquet files compared to `fetch_arrow_table()`.

#### To Reproduce

```
import pandas as pd
import duckdb
from pyarrow import parquet as pq

# Create dataset
n = 100_000
df = pd.DataFrame({"x": [0, 0, 1.1, 1.1, 2] * n, "y": ["A", "A", "A", "B", "B"] * n})
print(df.shape)

con = duckdb.connect()

# Query to arrow, then write to parquet -> 0.6 seconds
df_arrow = con.execute("SELECT * FROM df").fetch_arrow_table()
pq.write_table(df_arrow, "df_arrow.parquet")

# Copy to parquet -> 3.3 seconds
con.execute("COPY (SELECT * FROM df) TO 'df_duckdb.parquet' (FORMAT 'parquet')")

con.close()

# Check results
print(pd.read_parquet("df_duckdb.parquet").head())
print(pd.read_parquet("df_arrow.parquet").head())
```

The resulting file size:

![image](https://user-images.githubusercontent.com/16206095/160558700-4422ff19-0f96-44d0-afa0-14b4b949699f.png)

#### Environment (please complete the following information):
 - OS: Windows 10
 - DuckDB Version: 0.3.2
 - DuckDB Client: Python

**Comment by hannes** [comment]
This is probably due to DuckDB currently not emitting dictionary data pages in Parquet.

**Comment by lorentzenchr** [comment]
@hannes Thanks for your prompt reply. Is there anything someone like me (newbie but big fan on duckdb) could help with?

**Comment by hannes** [comment]
@lorentzenchr certainly, we are happy to review a PR that adds dictionary page support to DuckDB's parquet writer!

**Comment by lorentzenchr** [comment]
If it's within reach for a non-expert and you can provide some guidance, I can give it a try. Some pointer would certainly help.

**Comment by lorentzenchr** [comment]
🤔 https://arrow.apache.org/blog/2019/09/05/faster-strings-cpp-parquet/
> One of the largest and most complex optimizations involves encoding and decoding Parquet files’ internal dictionary-encoded data streams to and from Arrow’s in-memory dictionary-encoded DictionaryArray representation.

**Comment by mayer79** [comment]
@hannes, I would love to see this part being implemented, but based on @lorentzenchr's comment, it seems out of scope for an "ousider".

**Comment by ak2k** [comment]
Until this is implemented in duckdb, perhaps a work-around might be to use pyarrow's parquet writer ([via the duckdb-arrow integration](https://duckdb.org/2021/12/03/duck-arrow.html)), which seems to support dictionary compression.

**Comment by mayer79** [comment]
Thanks, @ak2k . This is indeed a good idea. If the arrow table does not fit into RAM: Is it possible to do this in a chunkwise fashion? I could not find a good example on how to do this. So any further hint is highly appreciated.

**Comment by ak2k** [comment]
Regrettably, I don't yet have experience with this; however, on the writing side, these two leads from the documentation might be helpful: 
- Writing a Parquet file with multiple row groups: https://arrow.apache.org/docs/python/parquet.html#finer-grained-reading-and-writing
- Writing a dataset with record batches: https://arrow.apache.org/docs/python/dataset.html#writing-large-amounts-of-data

**Comment by lorentzenchr** [comment]
With duckdb 0.5, this got a little better (~factor of 2):
```bash
% ls -lh df*
... 9.6K Sep 17 19:09 df_arrow.parquet
... 204K Sep 17 19:09 df_duckdb.parquet
```

**Comment by marklit** [comment]
Is this issue still present in 0.7? I produced a PQ file in https://github.com/duckdb/duckdb/discussions/6478 with both DuckDB 0.7 and ClickHouse and they were within 10% of one another in size. I've also produced the same file with a number of [Rust and Python](https://tech.marksblogg.com/postgresql-to-bigquery.html) libraries with similar file sizes.

**Comment by mayer79** [comment]
Rerunning above code with version 0.7.1 clearly shows some improvement, but there is still a factor of 20:

<img width="520" alt="image" src="https://user-images.githubusercontent.com/16206095/224477692-460ec372-0047-4941-a825-10889501ba30.png">

Maybe your dataset does not contain compressable columns? The difference comes from the character column.

**Comment by Mytherin** [comment]
We support dictionary encoding now, which de-duplicates strings within row groups. However, we don't yet support the various integer compression methods (such as the RLE-bitpacking hybrid) to compress the dictionary offsets. This has a big effect in this specific measurement because all of the strings are 1 byte, meaning the bulk of the data is actually not the strings itself, but the offsets. In most real world data sets the actual strings are longer, and hence supporting this compression has a much lower effect (on the order of 5-10% sounds right). Nevertheless we should still add support for it eventually.

**Comment by mayer79** [comment]
Thanks for the explanations @Mytherin .

A random double column and a random string solumn shows almost equal parquet size (using duckdb 0.7.1):

```
import numpy as np
import pandas as pd
import duckdb

# Create dataset
n = 1_000_000
str_1 = "This is a quite long string"
str_2 = "This string is even a little bit longer"

df = pd.DataFrame(
    {
        "x": np.random.uniform(size=n), 
        "y": np.random.choice([str_1, str_2], size=n)
    }
)
        
print(df.shape)

con = duckdb.connect()

# Write to parquet
df.to_parquet("df_pyarrow.parquet")
con.execute("COPY (SELECT * FROM df) TO 'df_duckdb.parquet' (FORMAT 'parquet')")

con.close()

# Check results
print(pd.read_parquet("df_duckdb.parquet").head())
print(pd.read_parquet("df_pyarrow.parquet").head())
```

<img width="510" alt="image" src="https://user-images.githubusercontent.com/16206095/224533586-d1d3b240-187e-49d3-b042-12222dcf1973.png">

**However**, when the numbers are also sampled from a discrete distribution, again we see a large difference. Is this related to the offsets?

```
# Create dataset
n = 1_000_000
str_1 = "This is a quite long string"
str_2 = "This string is even a little bit longer"

df = pd.DataFrame(
    {
        "x": np.random.choice([1.1, 400000, 0.45681], size=n), 
        "y": np.random.choice([str_1, str_2], size=n)
    }
)
```

<img width="524" alt="image" src="https://user-images.githubusercontent.com/16206095/224533532-dbd18577-4df3-4c64-92fe-50dabf3959fc.png">

**Comment by mayer79** [comment]
Since it seems to be a standard workflow to store the result as parquet, I'd love to keep the issue open.

**Comment by mayer79** [comment]
Same comment as last time

**Comment by mfouesneau** [comment]
Is there any more understanding on this issue? 
I find the serialized size is about a factor of 3 bigger for duckdb

**Comment by szarnyasg** [comment]
@mfouesneau yes, see https://github.com/duckdb/duckdb/issues/10108#issuecomment-1903664904. However, adding more advanced compression techniques for Parquet files is not planned for the next release.

**Comment by mfouesneau** [comment]
How does #10108 give more information? It is closed and refers to this current one.

**Comment by szarnyasg** [comment]
Ah I see that the comment I linked is basically saying the same that https://github.com/duckdb/duckdb/issues/3316#issuecomment-1464906646. So no new information in that thread then.

The only information I can add is in my previous comment: we know Parquet is important and we'll tackle this eventually but not in v0.10.

**Comment by lorentzenchr** [comment]
It‘s still an unresolved issue.

**Comment by mayer79** [comment]
Still very relevant!

**Comment by Mytherin** [comment]
There's work being done on adding support for different encodings to the Parquet writer - see https://github.com/duckdb/duckdb/pull/14257

**Comment by mayer79** [comment]
Much appreciated!

**Comment by gaborcsardi** [comment]
I have also ran into this. It seems to me that when the RLE/BP encoder is writing out dictionary indices it does not do any bit-packing. At least for me, the data page is either written in PLAIN encoding, or it is written as RLE_DICTIONARY, but only with repeats, no bit-packed parts at all. 

Which is quite wasteful if there is little or no repetition in dictionary encoded columns. Snappy or other compression helps a bit, but e.g. if I write `nycflights13::flights` to Parquet in Duckdb, I get a file that is twice as large as with Arrow or nanoparquet.

Some experiments in R:
```r
# function to write a data frame to a Parquet file with DuckDB
write_duckdb <- function(x, path) {
  drv <- duckdb::duckdb()
  con <- DBI::dbConnect(drv)
  on.exit(DBI::dbDisconnect(con), add = TRUE)
  DBI::dbWriteTable(con, "tab", as.data.frame(x))
  DBI::dbExecute(con, DBI::sqlInterpolate(con,
    "COPY tab TO ?filename (FORMAT 'parquet', COMPRESSION 'snappy')",
    filename = path
  ))
}
```

```r
# simple data that reproduces the issue
set.seed(1)
d <- data.frame(n = rep(sample(letters[1:3], 1000, replace = TRUE)))
d$n[1] <- NA   # to force an OPTIONAL column in nanoparquet
write_duckdb(d, "/tmp/ddb.parquet")
nanoparquet::write_parquet(d, "/tmp/np.parquet")

file.size("/tmp/ddb.parquet")
# [1] 873
file.size("/tmp/np.parquet")
# [1] 651
```

(This is actually not _that_ bad, because snappy compresses it nicely, real data was worse for me.)

It is dictionary encoded, so that's good:
```r
nanoparquet:::read_parquet_pages("/tmp/ddb.parquet")
#> # A data frame: 2 × 14
#>   file_name row_group column page_type page_header_offset uncompressed_page_size
#>   <chr>         <int>  <int> <chr>                  <dbl>                  <int>
#> 1 /tmp/ddb…         0      0 DICTIONA…                  4                     15
#> 2 /tmp/ddb…         0      0 DATA_PAGE                 34                   1312
#> # ℹ 8 more variables: compressed_page_size <int>, crc <int>, num_values <int>,
#> #   encoding <chr>, definition_level_encoding <chr>,
#> #   repetition_level_encoding <chr>, data_offset <dbl>,
#> #   page_header_length <int>
```

If you look at the actual data page:
```r
nanoparquet:::read_parquet_page("/tmp/ddb.parquet", 34L)$data[1:100]
#>   [1] 05 00 00 00 02 00 ce 0f 01 02 02 00 02 01 02 02 02 01 04 00 04 02 04 00 06
#>  [26] 01 08 02 02 00 02 01 02 00 08 01 02 02 04 01 06 02 02 01 02 00 02 01 02 00
#>  [51] 08 02 02 00 02 02 02 01 02 00 02 02 04 01 02 00 04 02 04 00 08 02 02 01 08
#>  [76] 02 02 01 04 00 02 02 04 00 02 02 04 00 08 01 02 00 02 02 02 00 04 01 02 02
```

The initial part is the definition levels, skip the first 10 bytes. Then the data starts:
```
02 00 02 01 02 02 02 01 04 00 04 02 04 00 06 01
...
```
Which is all RLE-runs, `02 00` is a length-1 0, `02 01` is a length-1 1, `02 02` is a length-1 2, `04 00` is a length-2 0, etc.

Ideally, the encoder would look ahead a bit to see if there are enough repetitions for an RLE-run, and if not it would do a bit-packed run. E.g. for bit-width 2 you'd need (about?) 6 repeated values, etc.

**Comment by Mytherin** [comment]
Have you tried this using the latest pre-release version? This should now be addressed, especially when writing v2 pages (`PARQUET_VERSION 'v2'`).

**Comment by gaborcsardi** [comment]
I get the same with the nightly build of the R package (Packaged: 2025-01-24), and that build does not know `PARQUET_VERSION`. The R package probably does not have the latest dev, though.

**Comment by Mytherin** [comment]
I get the following file-sizes for `nycflights13::flights`, so still larger than nanoparquet but improved from 10MB to 8MB in `v1.2.0` (when using `PARQUET_VERSION 'v2'`).

```sql
-rw-r--r--  1 myth  wheel   5.5M Jan 28 11:47 /tmp/np.parquet
-rw-r--r--  1 myth  wheel   9.3M Jan 28 11:47 /tmp/ddb-duckdb-v1.2.0-parquet-v1.parquet
-rw-r--r--  1 myth  wheel   8.0M Jan 28 11:47 /tmp/ddb-duckdb-v1.2.0-parquet-v2.parquet
-rw-r--r--  1 myth  wheel    10M Jan 28 11:47 /tmp/ddb-duckdb-v1.1.3.parquet
```

**Comment by lorentzenchr** [comment]
With duckdb-1.1.4.dev5024, I get for the OP's code snippet:
```
ls -lh df*
-rw-r--r--  1 lorentzen  staff   9.6K Jan 28 20:14 df_arrow.parquet
-rw-r--r--  1 lorentzen  staff    48K Jan 28 20:14 df_duckdb.parquet
 ```
That's a 10x better compression than with duckdb 0.3.2. So :rocket: quite some progress here.

**Comment by mayer79** [comment]
According to @lorentzenchr , one would need to set, see https://duckdb.org/2025/01/22/parquet-encodings.html

```
con.execute("COPY (SELECT * FROM df) TO 'df_duckdb.parquet' (PARQUET_VERSION V2)")
```

**Comment by Mytherin** [comment]
This is the current state - so this issue is mostly fixed.

```
$ > ls -trlah df_*.parquet
-rw-r--r--  1 myth  staff   9.5K Mar 20 08:34 df_arrow.parquet
-rw-r--r--  1 myth  staff    14K Mar 20 08:34 df_duckdb.parquet
```
